File: problems-originals/1601-1700/1625_lexicographically-smallest-string-after-applying-operations/solution.go
func findLexSmallestString(s string, a int, b int) string {
	n := len(s)
	seen := map[string]bool{s: true}
	queue := []string{s}
	best := s

	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]
		if cur < best {
			best = cur
		}

		digits := []byte(cur)
		for i := 1; i < n; i += 2 {
			value := (int(digits[i]-'0') + a) % 10
			digits[i] = byte('0' + value)
		}
		added := string(digits)
		if !seen[added] {
			seen[added] = true
			queue = append(queue, added)
		}

		rotated := cur[n-b:] + cur[:n-b]
		if !seen[rotated] {
			seen[rotated] = true
			queue = append(queue, rotated)
		}
	}

	return best
}

File: problems-originals/1601-1700/1625_lexicographically-smallest-string-after-applying-operations/solution.py
from collections import deque
from typing import List, Optional


class Solution:
    def findLexSmallestString(self, s: str, a: int, b: int) -> str:
        n = len(s)
        seen = {s}
        queue = deque([s])
        best = s

        while queue:
            cur = queue.popleft()
            if cur < best:
                best = cur

            digits = list(cur)
            for i in range(1, n, 2):
                digits[i] = str((int(digits[i]) + a) % 10)
            added = "".join(digits)
            if added not in seen:
                seen.add(added)
                queue.append(added)

            rotated = cur[-b:] + cur[:-b]
            if rotated not in seen:
                seen.add(rotated)
                queue.append(rotated)

        return best
